### PR TITLE
impl(GCS+gRPC): set timeouts on async downloads

### LINKED ISSUE #13453: Set timeouts on downloads
Or more precisely, use the progress timeouts via https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/internal/async_streaming_read_rpc_timeout.h

### PR DESCRIPTION
**Pull request by coryan**
Fixes #13453

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13801)

### CODE REVIEW

File: google/cloud/storage/internal/async/connection_impl.cc [hunk 1]
```diff
@@ -297,8 +302,14 @@ AsyncReaderConnectionFactory AsyncConnectionImpl::MakeReaderConnectionFactory(
                       google::storage::v2::ReadObjectRequest const& proto)
       -> future<StatusOr<std::unique_ptr<StreamingRpc>>> {
     ApplyQueryParameters(*context, *options, request);
+    auto timeout = ScaleStallTimeout(
+        options->get<storage::DownloadStallTimeoutOption>(),
+        options->get<storage::DownloadStallMinimumRateOption>(),
```
Comment (devbww) [inline google/cloud/storage/internal/async/connection_impl.cc:307]: Slightly interesting sub-text:  While `Options::get<>()` guarantees that you at least get a value-initialized duration, it is unclear (to me) that a zero "Rep" guarantees a zero duration.  I imagine it always does in practice though.

File: google/cloud/storage/internal/async/connection_impl_test.cc [hunk 2]
```diff
@@ -399,6 +420,115 @@ TEST_F(AsyncConnectionImplTest, AsyncReadObject) {
...
+    EXPECT_CALL(*stream, Finish).WillOnce([&] {
+      return sequencer.PushBack("Finish").then([](auto) { return Status{}; });
+    });
+    return std::unique_ptr<AsyncReadObjectStream>(std::move(stream));
+  });
+
+  auto mock_cq = std::make_shared<MockCompletionQueueImpl>();
+  // We will configure the connection to use 1 second timeouts.
+  EXPECT_CALL(*mock_cq, MakeRelativeTimer(
+                            std::chrono::nanoseconds(std::chrono::seconds(1))))
```
Comment (devbww) [inline google/cloud/storage/internal/async/connection_impl_test.cc:459]: We don't need to cast to nanoseconds first, right?  `MakeRelativeTimer()` will do it for us.

File: google/cloud/storage/internal/async/connection_impl_test.cc [hunk 3]
```diff
@@ -399,6 +420,115 @@ TEST_F(AsyncConnectionImplTest, AsyncReadObject) {
...
+  ASSERT_TRUE(pending.is_ready());
+  auto r = pending.get();
+  ASSERT_STATUS_OK(r);
+  auto reader = *std::move(r);
+
+  // First simulate a failed `ReadObject()`. This returns a streaming RPC
+  // that completes with `false` on `Start()` (i.e. never starts) and then
+  // completes with a transient error on `Finish()`. Because the timeout
+  // parameters are configured, the first thing to happen is that the timeout is
+  // set:
```
Comment (devbww) [inline google/cloud/storage/internal/async/connection_impl_test.cc:491]: s/:/./

File: google/cloud/storage/internal/async/connection_impl_test.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/async/connection_impl_test.cc:459]: This is setting a mock expectation, and `std::chrono::nanoseconds(std::chrono::seconds(1))` is just the matcher.  If I use `std::chrono::seconds()` I get:

```
[build] /usr/local/google/home/coryan/google-cloud-cpp/google/cloud/storage/internal/async/connection_impl_test.cc:459:42: error: no matching function for call to ‘google::cloud::v2_23::testing_util::MockCompletionQueueImpl::gmock_MakeRelativeTimer(std::chrono::seconds)’
[build]   459 |   EXPECT_CALL(*mock_cq, MakeRelativeTimer(
[build]       |                                          ^
```

I guess if we had a matcher that could compare `std::chrono::duration` with different `Period` parameters I could use that instead, but we don't.

Comment (coryan) [reply google/cloud/storage/internal/async/connection_impl_test.cc:459]: I tried with `Eq(std::chrono::seconds(1))`, no luck either 🤷

File: google/cloud/storage/internal/async/connection_impl.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/async/connection_impl.cc:307]: Ack.

File: google/cloud/storage/internal/async/connection_impl_test.cc [hunk 4]
```diff
@@ -399,6 +420,115 @@ TEST_F(AsyncConnectionImplTest, AsyncReadObject) {
...
+  EXPECT_EQ(next.second, "Read");
+  timer.first.set_value(false);  // simulate a canceled timeout
+  next.first.set_value(true);
+
+  auto response = data.get();
+  ASSERT_TRUE(
+      absl::holds_alternative<storage_experimental::ReadPayload>(response));
+
+  // Trigger another read. Since this closes the stream, the `Read()` and
+  // `Finish()` calls must happen before the second `Read()` it is satisfied.
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/async/connection_impl_test.cc]: Fixed

File: google/cloud/storage/internal/async/connection_impl_test.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/internal/async/connection_impl_test.cc]: Fixed

File: google/cloud/storage/internal/async/connection_impl_test.cc (hunk 2 above)
Comment (devbww) [reply google/cloud/storage/internal/async/connection_impl_test.cc:460]: I see.  Thanks.

**Pull request merged**